### ISSUE TITLE
x64 build: file open performance

### ISSUE THREAD

**Issue by Rikk** [notepad-plus-plus/notepad-plus-plus#668]
So far, my x64 build looks cute, but.... even if it is able to load very large files, it is very slow for the task. In other words, it takes long time (will full cpu) to completely open -- until it allows scroll till the end -- files of hundreds megabytes. *Edit: looks related to "Word Wrap" feature, so, turn it off when opening those files*.

Is there anyone working on file open performance for x64 (or maybe x86 too)?
Any idea on how to improve it?

[For those needing a x64 build, I have it [available here](http://portfolio.ricardo.solutions/etc/Npp-x64.bin.7z). Note that it is not compatible with 32-bit plugins (ie. all available plugins), and may be unstable for some tasks (like opening very large files *with "Word Wrap" on*).]

**Comment by ariccio** [comment]
[Profile it.](http://blogs.msdn.com/b/visualstudioalm/archive/2014/02/28/new-cpu-usage-tool-in-the-performance-and-diagnostics-hub-in-visual-studio-2013.aspx)

**Comment by milipili** [comment]
If the x64 build is slow, it may indicate several issues in the code, and should be reproductible in 32 bits (maybe more difficult to reproduce). The file loading will be rewritten but it could be interresting to know by advance where the difficulties are.

**Comment by Rikk** [comment]
As I edited above, it seems main problem is with Line Wrap, even if no line is actually wrapping in the test file.
I'll be profiling it later, as well suggested, thanks ariccio.

**Comment by Rikk** [comment]
Was so curious to see what was that, so I have this:
With "Word Wrap" on:
![slow](https://cloud.githubusercontent.com/assets/2916485/9132983/0e863ecc-3cce-11e5-8804-782b43f8573e.png)
The "cpu" hungry appointed one is "[external code]", but previous function is the `ScintillaEditView::scintillaNew_Proc`:
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/ScitillaComponent/ScintillaEditView.cpp#L298

@ariccio how we can find the problem? Is that the culprit function?

"Word Wrap" off (I see no big problem):
![less-slow](https://cloud.githubusercontent.com/assets/2916485/9133090/149f7b92-3ccf-11e5-8e59-8f9bb4ce825b.png)

**Comment by milipili** [comment]
The issue may not come from this function exactly, but is probably a consequence. The word wrap mode may force to send a lot of windows message.
It should be reproductible in x86 so I will try later.

**Comment by milipili** [comment]
Before I try something : is the behavior similar with VS2013 ?

**Comment by Rikk** [comment]
sorry, don't have it installed. (I don't intend to install it either (that is truly a time-consuming task lol)).

**Comment by Rikk** [comment]
One thing you could look at is this [I posted in community](https://notepad-plus-plus.org/community/topic/237/editing-600-mega-xml-file/10):
If I have no plugins enabled and no other file opened, I can open files with maximum size around 362MB, before it displays “File is too big to be opened” error.
But with my plugins enabled, the limit lowers to about 150MB.

Both cases are much lower than the limit of RAM a 32-bit program can allocate (~2.5-3GB).
Btw, the x64 build I made used around 2GB (with spike to 3GB) to open let's say a 900MB file.

**Comment by milipili** [comment]
Even for a 32 bit program, a 600 MiB file should be loaded easily (unless there is something tricky I don't know). It is for me a bug that should be reported.

**Comment by ariccio** [comment]
@Rikk It looks like `Just my code` was turned on - it is by default. I believe that the checkbox is in `Filter view`. If you `Create` [a] `detailed report`, you'll get more information.

**Comment by Rikk** [comment]
Okay, looks like it's up to Scintilla `PositionCache::MeasureWidths` :arrow_right: `SurfaceGDI::MeasureWidths` :arrow_right: *gdi32.dll* :arrow_right: ???

This round I tested Release x86 with `<GenerateDebugInformation>true</GenerateDebugInformation>` (yea, ariccio, it is useful for something :smile:)

Can these .vsps (analyzed reports?) files be viewed in a machine other than the one that created them? If so, I share [mine here](http://portfolio.ricardo.solutions/etc/Report20150808-0201.vsps).

**Comment by Rikk** [comment]
Nowadays, shouldn't it use DirectWrite to render? Afaik, gdi32 is a thing from win 9x era.

**Comment by ariccio** [comment]
> gdi32.dll :arrow_right: ???

Are your symbols configured correctly?

In short:  https://support.microsoft.com/en-us/kb/311503

In depth: https://www.osr.com/blog/2014/10/01/setting-the-windbg-symbol-search-path/

> Nowadays, shouldn't it use DirectWrite to render? Afaik, gdi32 is a thing from win 9x era.

Ideally, yea. But it's not *that bad*, and DirectWrite is a bit more work. This is partly for compatibility reasons, as DirectWrite (and it's other high-performance family members, like DirectComposition) isn't supported under Windows XP. We might want to come back to this in the future.

**Comment by ariccio** [comment]
> Are your symbols configured correctly?

Ahh, nevermind. I'm hitting the same problem... and yes, the analyzed reports are sharable. They're also apparently backwards (2015-2013) compatible.


I might have to get my vTune license straightened out.

**Comment by ariccio** [comment]
(Sorry to triple post!)

> I might have to get my vTune license straightened out.

Errr... nevermind. I have a headache right now, and it's not helping me think this through (D'oh!)

The hot path is pretty clear:

![measurewidths_hotpath](https://cloud.githubusercontent.com/assets/2142308/9149597/6e100772-3d7f-11e5-9828-4b72ea625e67.PNG)

**Comment by Rikk** [comment]
Exactly, now your turn to find the wrong code there xD

**Comment by Rikk** [comment]
Then maybe we should make a pull for scintilla if this improves our slowness -- or keep it for our mod, if it breaks any compat they may want to have with 9x cr*p.

**Comment by Rikk** [comment]
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/scintilla/win32/ScintillaWin.cxx#L34
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/scintilla/win32/PlatWin.cxx#L29

cya later

**Comment by Rikk** [comment]
After above comment I saw Scintilla has an alternative class `SurfaceD2D` for these commands, instead of `SurfaceGDI`. Problem is that it checks if NTDDI_WIN7 is defined, I suppose it means it will be enabled only in Win7, right?

https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/scintilla/win32/ScintillaWin.cxx#L34
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/scintilla/win32/PlatWin.cxx#L29

Anyway, when I compiled bypassing that check, Npp crashed right at initialization when using these directx functions. Haven't debugged yet, as today I was looking at other things.
